### ISSUE TITLE
core: consider `Value` implementation for `Option<T> where T: Value`

### ISSUE THREAD

**Issue by hawkw** [tokio-rs/tracing#824]
### Crates

- `tracing-core`

### Motivation

Currently, if I want to record an `Option` as a field value, I have to do one of the following:

* record the `Option` using its `fmt::Debug` impl:
  ```rust
  let foo: Option<&str> = ...;
  tracing::info!(?foo, bar);
  ```
* emit two separate events based on whether or not the option is present:
  ```rust
  if let Some(ref foo) = foo {
      tracing::info!(?foo, bar);
  } else {
      tracing::info!(bar);
  }
  ```

Neither of these approaches are ideal. When using `fmt::Debug`, the type inside the `Option` is _not_ recorded as a typed value --- even if it does have a value implementation. So, if I wanted to record an `Option<u64>` this way, the subscriber would not be able to consume it as a number, only as `dyn Debug`. Similarly, it would be hard to record an `Option` where the optional type is `Display` using its `Display` implementation, rather than `Debug`.

The approach where two separate events are emitted is also not ideal (and is worse with spans). We've now created two separate callsites, on separate lines, for what is logically the same event. This means that if a subscriber is relying on callsites to track specific events that it cares about, these two events will not be considered the same. This isn't ideal.

### Proposal

The simplest, and perhaps nicest, approach would be to add an implementation of `Value` for `Option<T> where T: Value`. When the `Option` is `None`, the implementation could behave similarly to `field::Empty`, and simply perform a no-op:
https://github.com/tokio-rs/tracing/blob/96e9d6d4f0c01a6a260912f441ba17c9ea1eb734/tracing-core/src/field.rs#L517-L521

When the `Option` is `Some`, the inner value would be recorded using its `Value` impl.

By not recording optional values when they are not present, we end up with more concise output. Also, this would allow the way the inner value is formatted to be customized. For example, we could record a `&str` with its `Display` impl if it is present:
```rust
let foo: Option<&str> = ...;
tracing::info!(foo = foo.as_ref().map(tracing::field::display));
```

### Alternatives

The main alternative to having `Option` behave similarly to `Empty` when it is `None` is to add `Option` as a _new primitive type_. This would mean adding
```rust
pub trait Visit {
    // ...
    fn record_option(&mut self, field: &Field, value: Option<&dyn Value>);
    // ...
}
```
to the `Visit` trait. Note that the optional value would have to be recorded as a `dyn Value` trait object, because `Visit` must be object-safe and therefore cannot have generic methods; the `Value` impl for `Option<T>` could handle the trait object coercion, though.

This is more similar to the approach used by `serde`, which treats `Option`s as a primitive and has [`serialize_some`](https://docs.rs/serde/1.0.114/serde/trait.Serializer.html#tymethod.serialize_some) and [`serialize_none`](https://docs.rs/serde/1.0.114/serde/trait.Serializer.html#tymethod.serialize_none) methods on its `Serializer` trait. Since we want to support compatibility with `serde` (see #819), it might be worth choosing the design that's closer to what `serde` does. 

Additionally, some serialization systems may distinguish between a field that _doesn't exist_ and a field that is _not present_. For example, suppose I have a JSON object that might contain `foo` and `bar` keys. If both exist, I would record 
```json
{
    "foo": 1,
    "bar": 2
}
```

If only `foo` exists, however, we might record
```json
{
    "foo": 1,
}
```
_or_
```json
{
    "foo": 1,
    "bar": null
}
```

I am not familiar enough with the ecosystem of JSON-parsing tools to know whether or not most systems consider these equivalent, or if there is a semantic difference. However, they definitely have a different _implications_ to a human reader: `"field": null` definitely suggests that there is _sometimes_ a value called  `"field"`, while `<empty string>` doesn't --- if I was only looking at a _single_ JSON object, I wouldn't have the context to be aware that that field sometimes exists.

Similarly, some users might prefer to have textual logs that explicitly print `field=None` rather than skipping the field --- I could then `grep` for `field=None` or use `tracing-subscriber`'s `EnvFilter` to see all the events where that value was `None`. An implementation of `Value` for `Option` that does the skipping internally wouldn't allow this.

**Comment by hawkw** [comment]
@davidbarsky @LucioFranco @anp any opinions?

**Comment by hawkw** [comment]
I just thought of a third option that's kind of a compromise between the two proposals I discussed above:

We add a `record_option` method to the `Visit` trait, and implement `Value` for `Option` by calling that method, as in the second proposal. However, when we add the default implementation for `record_option` (which we need to do, because otherwise, adding a method without a default is a breaking change), rather than having it do the usual behavior for `record_$TY` methods (falling back to `self.record_debug`), we would instead implement the "skip recording if `None` behavior of the first proposal.

This way, users get the more concise behavior out of the box without having to think about it, and can treat `Option`s as native values freely. But, unlike skipping `None`s in the `Value` impl for `Option`s, an implementation of `Visit` for a `serde` serializer can dispatch to `serde`'s `serialize_some` and `serialize_none` methods. Similarly, a user who wants to see `field=None` in textual logs can override the default field formatter with that behavior.

IMO, this is the right approach, but I'd love to hear what others think.

**Comment by anp** [comment]
This SGTM! We'll probably skip encoding optional fields altogether once we have a way to attach schemas to callsites, since a renderer is where we'd want to make the None explicit in order to save on bytes.

**Comment by LucioFranco** [comment]
I like that third approach :+1:

**Comment by hawkw** [comment]
Okay, I think the third option is probably the right choice. To recap:

1. We want to add a `record_option` method to the `Visit` trait in `tracing_core::field`, like:
  ```rust
  pub trait Visit {
      // ...
      fn record_option(&mut self, field: &Field, value: Option<&dyn Value>);
      // ...
  }
  ```
2. We add a default implementation of that method that matches the value, returns immediately if it is `None`, and calls `Value::record` on the inner value if it is `Some`, passing `self` as the visitor.
3. We add an implementation of `Value` for `Option<T> where T: Value` that calls `Visit::record_option`, converting the `&Option<T>` to an `Option<&dyn Value>`. We would probably do that with code along the lines of 
  ```rust
  self.as_ref().map(|v| v as &dyn Value)
  ```

I'm tagging this as a "good first issue", since I think the implementation should be pretty straightforward now that we've agreed on an approach.

**Comment by humancalico** [comment]
I'll try to work on this

**Comment by hawkw** [comment]
> I'll try to work on this

@humancalico Great, let me know if you have any questions!

**Comment by Pothulapati** [comment]
@humancalico Are you still working on the issue? I'm happy to take this up, if you are not! :)

**Comment by humancalico** [comment]
@Pothulapati I am

**Comment by hawkw** [comment]
Closed by #1585.
